File: go/ppc/metasyntax.go
package capitan

type HoleType int

const (
    Everything HoleType = iota
    Expression
    Alphanum
    NonSpace
    Line
    Blank
)

type Delimiter struct {
    Start *string
    End   *string
}

type ReservedIdentifiers []string

type Hole struct {
    Type      HoleType
    Delimiter interface{}
}

type Regex struct {
    Start string
    Sep   byte
    End   string
}

type Syntax []interface{}

type Alias struct {
    Pattern       string
    MatchTemplate string
    Rule          *string
}

var defaultSyntax = Syntax{
    Regex{":[", '~', "]"},
    Hole{Everything, Delimiter{start: pstr(":["), end: pstr("]")}},
    Hole{Expression, Delimiter{start: pstr(":["), end: pstr(":e]")}},
    Hole{Alphanum, Delimiter{start: pstr(":[["), end: pstr("]]")}},
    Hole{NonSpace, Delimiter{start: pstr(":["), end: pstr(".]")}},
    Hole{Line, Delimiter{start: pstr(":["), end: pstr("\\n]")}},
    Hole{Blank, Delimiter{start: pstr(":[ "), end: pstr("]")}},
    Hole{Expression, ReservedIdentifiers{"α", "β", "γ", "δ", "ε", "ζ", "η", "θ", "ι", "κ", "λ", "μ", "ξ", "π", "ρ", "ς", "σ", "τ", "υ", "φ", "χ", "ψ", "ω"}},
    Hole{Everything, ReservedIdentifiers{"Γ", "Δ", "Θ", "Λ", "Ξ", "Π", "Σ", "Φ", "Ψ", "Ω"}},
}

var defaultIdentifier = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_"

var defaultAliases = []Alias{
    {Pattern: "...", MatchTemplate: ":[_]", Rule: nil},
}

var defaultMetasyntax = struct {
    Syntax     Syntax
    Identifier string
    Aliases    []Alias
}{
    Syntax:     defaultSyntax,
    Identifier: defaultIdentifier,
    Aliases:    defaultAliases,
}

func create(syntax Syntax, identifier string, aliases []Alias) struct {
    Syntax     Syntax
    Identifier string
    Aliases    []Alias
} {
    return struct {
        Syntax     Syntax
        Identifier string
        Aliases    []Alias
    }{
        Syntax:     syntax,
        Identifier: identifier,
        Aliases:    aliases,
    }
}

func Default() struct {
    Syntax     Syntax
    Identifier string
    Aliases    []Alias
} {
    return create(defaultSyntax, defaultIdentifier, defaultAliases)
}

func pstr(s string) *string {
    return &s
}

File: python/ppc/metasyntax.py
from dataclasses import dataclass
from enum import Enum, auto
from typing import Optional, Union, List

class HoleType(Enum):
    Everything = auto()
    Expression = auto()
    Alphanum = auto()
    NonSpace = auto()
    Line = auto()
    Blank = auto()

Delimited = Union[tuple[Optional[str], Optional[str]], None]
ReservedIdentifiers = List[str]
Delimiter = Union[Delimited, ReservedIdentifiers]

@dataclass
class Hole:
    hole_type: HoleType
    delimiter: Delimiter

@dataclass
class Regex:
    start: str
    sep: str
    end: str

Syntax = List[Union[Hole, Regex]]

@dataclass
class Alias:
    pattern: str
    match_template: str
    rule: Optional[str]

@dataclass
class Metasyntax:
    syntax: Syntax
    identifier: str
    aliases: List[Alias]

def default_syntax() -> Syntax:
    return [
        Regex(":[", '~', "]"),
        Hole(HoleType.Everything, (Some(":["), Some("]"))),
        Hole(HoleType.Expression, (Some(":["), Some(":e]"))),
        Hole(HoleType.Alphanum, (Some(":[["), Some("]]"))),
        Hole(HoleType.NonSpace, (Some(":["), Some(".]"))),
        Hole(HoleType.Line, (Some(":["), Some("\\n]"))),
        Hole(HoleType.Blank, (Some(":[ "), Some("]"))),
        Hole(HoleType.Expression, ["α", "β", "γ", "δ", "ε", "ζ", "η", "θ", "ι", "κ", "λ", "μ", "ξ", "π", "ρ", "ς", "σ", "τ", "υ", "φ", "χ", "ψ", "ω"]),
        Hole(HoleType.Everything, ["Γ", "Δ", "Θ", "Λ", "Ξ", "Π", "Σ", "Φ", "Ψ", "Ω"]),
    ]

def default_identifier() -> str:
    return "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_"

def default_aliases() -> List[Alias]:
    return [Alias("...", ":[_]", None)]

def default_metasyntax() -> Metasyntax:
    return Metasyntax(default_syntax(), default_identifier(), default_aliases())

def create(syntax: Syntax, identifier: str, aliases: List[Alias]) -> Metasyntax:
    return Metasyntax(syntax, identifier, aliases)

def default() -> Metasyntax:
    return create(default_syntax(), default_identifier(), default_aliases())